scrapy/pqueues.py

```diff
@@ -0,0 +1,218 @@
+import hashlib
+import logging
+from six import text_type
+from six.moves.urllib.parse import urlparse
+
+from queuelib import PriorityQueue
+
+from scrapy.core.downloader import Downloader
+from scrapy.http import Request
+from scrapy.signals import request_reached_downloader, response_downloaded
+
+
+logger = logging.getLogger(__name__)
+
+
+SCHEDULER_SLOT_META_KEY = Downloader.DOWNLOAD_SLOT
+
+
+def _get_from_request(request, key, default=None):
+    if isinstance(request, dict):
+        return request.get(key, default)
+
+    if isinstance(request, Request):
+        return getattr(request, key, default)
+
+    raise ValueError('Bad type of request "%s"' % (request.__class__, ))
+
+
+def _scheduler_slot_read(request, default=None):
+    meta = _get_from_request(request, 'meta', dict())
+    slot = meta.get(SCHEDULER_SLOT_META_KEY, default)
+    return slot
+
+
+def _scheduler_slot_write(request, slot):
+    meta = _get_from_request(request, 'meta', None)
+    if not isinstance(meta, dict):
+        raise ValueError('No meta attribute in %s' % (request, ))
+    meta[SCHEDULER_SLOT_META_KEY] = slot
+
+
+def _scheduler_slot(request):
+
+    slot = _scheduler_slot_read(request, None)
+    if slot is None:
+        url = _get_from_request(request, 'url')
+        slot = urlparse(url).hostname or ''
+        _scheduler_slot_write(request, slot)
+
+    return slot
+
+
+def _pathable(x):
+    pathable_slot = "".join([c if c.isalnum() or c in '-._' else '_' for c in x])
+
+    """
+        as we replace some letters we can get collision for different slots
+        add we add unique part
+    """
+    unique_slot = hashlib.md5(x.encode('utf8')).hexdigest()
+
+    return '-'.join([pathable_slot, unique_slot])
+
+
+class PrioritySlot:
```

Comment:
This looks like a tuple with a custom `__str__` method; do you think some code can be simplified if we make it just a tuple? For example, would it allow to remove PriorityAsTupleQueue?

Reply:
I will be glad if we can drop `PriorityAsTupleQueue`. Tuples encoded to json as lists, lists decoded from json as lists. Lists are unhashable. Only hashable types can be used inside common `PriorityQueue`